README.md
# HTTP REST server for Firebird Databases

A simple REST HTTP server that will parse FirebirdSQL query and return response in **JSON** format.

[![npm version](https://badge.fury.io/js/firebird-rest.svg)](https://badge.fury.io/js/firebird-rest)

## Installation

```
npm install -g firebird-rest
```

## Usage

```
Usage: firebird-rest [options]
```

## Flags

```
--port Port that app listens to (default: 4444)
--h    Firebird host (default: "localhost")
--p    Firebird port (default: 3050)
--db   Firebird database path (default:"")
--u    Firebird User (default: "SYSDBA")
--pw   User password (default: "masterkey")
--r    User role (default: null)
```

## Usage

### Port

This package is preconfigured to listen on **TCP port 4444**, however, you can specify any port you want with ``--port`` flag. If you do that, make sure you add that port to the list of accessible ports in firewall settings.  
For testing purposes you can use:

- [Postman](https://www.getpostman.com/)
- `curl` command from the console
- your browser by pointing it to:

```
http://localhost:<port>/?sql=<your sql query>
```

### Example

```
~ $ firebird --db /opt/firebird/examples/empbuild/employee.fdb
^Z
[1]+  Stopped firebird --db /opt/firebird/examples/empbuild/employee.fdb
~ $ bg
[1]+ firebird --db /opt/firebird/examples/empbuild/employee.fdb &
~ $ curl 'http://localhost:4444' -d sql='select * from country' -L
[{"COUNTRY":"USA","CURRENCY":"Dollar"},{"COUNTRY":"England","CURRENCY":"Pound"},{"COUNTRY":"Canada","CURRENCY":"CdnDlr"},{"COUNTRY":"Switzerland","CURRENCY":"SFranc"},{"COUNTRY":"Japan","CURRENCY":"Yen"},{"COUNTRY":"Italy","CURRENCY":"Lira"},{"COUNTRY":"France","CURRENCY":"FFranc"},{"COUNTRY":"Germany","CURRENCY":"D-Mark"},{"COUNTRY":"Australia","CURRENCY":"ADollar"},{"COUNTRY":"Hong Kong","CURRENCY":"HKDollar"},{"COUNTRY":"Netherlands","CURRENCY":"Guilder"},{"COUNTRY":"Belgium","CURRENCY":"BFranc"},{"COUNTRY":"Austria","CURRENCY":"Schilling"},{"COUNTRY":"Fiji","CURRENCY":"FDollar"}]
~ $ fg
firebird --db /opt/firebird/examples/empbuild/employee.fdb
^C
```

## Limitations

Charset for database connection is always **UTF-8**.
Node is unicode, no matter if your database is using another charset to store string or blob, Firebird will transliterate automatically.
This is why you should use Firebird 2.5 server at least.

### Firebird 3.0 Support

Firebird's new wire protocol is not supported yet, so for Firebird 3.0 you need to add the following in `firebird.conf`

```
AuthServer = Legacy_Auth
WireCrypt = Disabled
```

## License

[MIT](/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 1.0.0 - 2019-07-08
### Added
- New config.json file
- Option to use multiple databases with a single instance using the new config.json file
- New dependencies: node-windows and file-system
- Ability to install the program as a windows service by running the install-service.js file.

## 0.0.2 - 2014-07-10
### Added
- Option to specify server port with ``--port`` flag

## 0.0.1 - 2019-01-18
- Initial release


src/main.js
#!/usr/bin/env node

const express = require('express');
const compression = require('compression');
const bodyParser = require('body-parser');
const port1 = require('./utils/flagParams').serverPort1();
const activeDBs = require('./utils/flagParams').activeDBs();
const flagMode = require('./utils/flagParams').flagMode;
const sqlQuerys = require('./utils/sqlQuery')
const app1 = express();
const app2 = express();
const app3 = express();
const app4 = express();
const app5 = express();


app1.use(compression()); // attempt to compress response bodies
app1.use(bodyParser.urlencoded({
  extended: false,
  limit: '1mb'
})); // parsing the URL-encoded data with the ```querystring``` library
app1.use(bodyParser.json({
  limit: '1mb'
})); // maximum request body size
app1.use((req, res, next) => {
  req.connection.setTimeout(4 * 60 * 1000); // 4 minutes
  res.connection.setTimeout(4 * 60 * 1000); // 4 minutes
  next();
});

app1.get('/', sqlQuerys.sqlQuery1('query')); // parse SQL queries via http GET request
app1.post('/', sqlQuerys.sqlQuery1('body')); // parse SQL queries via http POST request

// RUN FIREBIRD SERVER
app1.listen(port1, () => console.log(`Database 1:\nFirebird Server up and running on\nhttp://localhost:${port1}`));

if (flagMode == "False") {
  if (activeDBs.db2 == "True") {
    const port2 = require('./utils/flagParams').serverPort2();
    app2.use(compression()); // attempt to compress response bodies
    app2.use(bodyParser.urlencoded({// parsing the URL-encoded data with the ```querystring``` library
      extended: false,
      limit: '1mb'
    }));
    app2.use(bodyParser.json({ // maximum request body size
      limit: '1mb'
    }));
    app2.use((req, res, next) => {
      req.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      res.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      next();
    });

    app2.get('/', sqlQuerys.sqlQuery2('query')); // parse SQL queries via http GET request
    app2.post('/', sqlQuerys.sqlQuery2('body')); // parse SQL queries via http POST request

    app2.listen(port2, () => console.log(`Database 2:\nFirebird Server up and running on\nhttp://localhost:${port2}`));
  }
  if (activeDBs.db3 == "True") {
    const port3 = require('./utils/flagParams').serverPort3();
    app3.use(compression()); // attempt to compress response bodies
    app3.use(bodyParser.urlencoded({
      extended: false,
      limit: '1mb'
    })); // parsing the URL-encoded data with the ```querystring``` library
    app3.use(bodyParser.json({
      limit: '1mb'
    })); // maximum request body size
    app3.use((req, res, next) => {
      req.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      res.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      next();
    });

    app3.get('/', sqlQuerys.sqlQuery3('query')); // parse SQL queries via http GET request
    app3.post('/', sqlQuerys.sqlQuery3('body')); // parse SQL queries via http POST request

    app3.listen(port3, () => console.log(`Database 3:\nFirebird Server up and running on\nhttp://localhost:${port3}`));
  }
  if (activeDBs.db4 == "True") {
    const port4 = require('./utils/flagParams').serverPort4();
    app4.use(compression()); // attempt to compress response bodies
    app4.use(bodyParser.urlencoded({
      extended: false,
      limit: '1mb'
    })); // parsing the URL-encoded data with the ```querystring``` library
    app4.use(bodyParser.json({
      limit: '1mb'
    })); // maximum request body size
    app4.use((req, res, next) => {
      req.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      res.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      next();
    });

    app4.get('/', sqlQuerys.sqlQuery4('query')); // parse SQL queries via http GET request
    app4.post('/', sqlQuerys.sqlQuery4('body')); // parse SQL queries via http POST request

    app4.listen(port4, () => console.log(`Database 4:\nFirebird Server up and running on\nhttp://localhost:${port4}`));
  }
  if (activeDBs.db5 == "True") {
    const port5 = require('./utils/flagParams').serverPort5();
    app5.use(compression()); // attempt to compress response bodies
    app5.use(bodyParser.urlencoded({
      extended: false,
      limit: '1mb'
    })); // parsing the URL-encoded data with the ```querystring``` library
    app5.use(bodyParser.json({
      limit: '1mb'
    })); // maximum request body size
    app5.use((req, res, next) => {
      req.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      res.connection.setTimeout(4 * 60 * 1000); // 4 minutes
      next();
    });

    app5.get('/', sqlQuerys.sqlQuery5('query')); // parse SQL queries via http GET request
    app5.post('/', sqlQuerys.sqlQuery5('body')); // parse SQL queries via http POST request

    app5.listen(port5, () => console.log(`Database 5:\nFirebird Server up and running on\nhttp://localhost:${port5}`));
  }

}


src/utils/sqlQuery.js
const Firebird = require('node-firebird-dev');
const Options = require('./flagParams').options();
const convertBufferArray = require('./convertBufferArray');

const sqlQuery = param => {
  return (req, res) => {
    let result = [];
    const properties = req[param];
    
    Options.database = properties.database || Options.database;
    Options.user = properties.user || Options.user;
    Options.password = properties.password || Options.password;
    Options.role = properties.role || Options.role;

    const sql = properties.sql;
    const params = properties.params;

    if (!sql) {
      return res.send(['No valid SQL query found! Please enter a valid SQL query.']);
    }

    Firebird.attach(Options, (err, db) => {
      if (err) {
        db.detach();
        res.status(400); // BAD REQUEST
        return res.send(`\n${err.message}\n`);
      }

      db.query(sql, params, (err, data) => {
        if (err) {
          db.detach();
          res.status(400); // BAD REQUEST
          return res.send(`\n${err.message}\n`);
        }
        db.detach();

        if (data) {
          if (Array.isArray(data)) {
            // CONVERT RAW QUERY RESULT AND RETURN JSON
            data.forEach(row => {
              const newRow = convertRow(row);
              result.push(newRow);
            });
          } else {
            const newRow = convertRow(data);
            result = newRow;
          }
        }

        return res.send(result);
      });
    });
  };
};

function convertRow(row) {
  let newRow = {};
  Object.keys(row).forEach(el => {
    newRow[el] = convertBufferArray(row[el]);
  });

  return newRow;
}

module.exports = sqlQuery;


src/utils/flagParams.js
const flags = require('flags');
var fs = require("file-system");
const path = require('path');
var config = fs.readFileSync(path.resolve(require('path').dirname(require.main.filename) + "/utils","./config.json"));
var jsonConfig = JSON.parse(config);

flags.defineNumber('port', '4444', 'Port that app listens to');
flags.defineString('h', 'localhost', 'Firebird Host');
flags.defineNumber('p', '3050', 'Firebird Port');
flags.defineString('db', 'NO_DATABASE', 'Absolute path to Firebird Database');
flags.defineString('u', 'SYSDBA', 'Firebird User');
flags.defineString('pw', 'masterkey', 'Firebird User Password');
flags.defineString('r', null, 'Firebird User Role');

flags.parse();

var flagMode = "False";
var activeDBs;
var options1;
var options2;
var options3;
var options4;
var options5;
var serverPort1;
var serverPort2;
var serverPort3;
var serverPort4;
var serverPort5;

//If the flag mode is active, fill options1 with flags
if (flags.get('db') != 'NO_DATABASE') {
  flagMode = "True";
  options1 = () => {
    return {
      host: flags.get('h'),
      port: flags.get('p'),
      database: flags.get('db'),
      user: flags.get('u'),
      password: flags.get('pw'),
      role: flags.get('r')
    };
  };
  serverPort1 = () => {
    return flags.get('port');
  };
  activeDBs = () => {
    return {
      db1: "True",
      db2: "False",
      db3: "False",
      db4: "False",
      db5: "False"
    }
  }
}
//If the flag mode is not active fill options 1 - 5 with config.json params
else {
  //Database 1.
  if (jsonConfig.options1.active == "True") {
    options1 = () => {
      return {
        host: jsonConfig.options1.host,
        port: jsonConfig.options1.port,
        database: jsonConfig.options1.database,
        user: jsonConfig.options1.user,
        password: jsonConfig.options1.password,
        role: jsonConfig.options1.role
      };
    };
    serverPort1 = () => {
      return jsonConfig.options1.srvPort;
    };
  }
  //Database 2.
  if (jsonConfig.options2.active == "True") {
    options2 = () => {
      return {
        host: jsonConfig.options2.host,
        port: jsonConfig.options2.port,
        database: jsonConfig.options2.database,
        user: jsonConfig.options2.user,
        password: jsonConfig.options2.password,
        role: jsonConfig.options2.role
      };
    };
    serverPort2 = () => {
      return jsonConfig.options2.srvPort;
    };
  }
  //Database 3.
  if (jsonConfig.options3.active == "True") {
    options3 = () => {
      return {
        host: jsonConfig.options3.host,
        port: jsonConfig.options3.port,
        database: jsonConfig.options3.database,
        user: jsonConfig.options3.user,
        password: jsonConfig.options3.password,
        role: jsonConfig.options3.role
      };
    };
    serverPort3 = () => {
      return jsonConfig.options3.srvPort;
    };
  }
  //Database 4.
  if (jsonConfig.options4.active == "True") {
    options4 = () => {
      return {
        host: jsonConfig.options4.host,
        port: jsonConfig.options4.port,
        database: jsonConfig.options4.database,
        user: jsonConfig.options4.user,
        password: jsonConfig.options4.password,
        role: jsonConfig.options4.role
      };
    };
    serverPort4 = () => {
      return jsonConfig.options4.srvPort;
    };
  }
  //Database 5.
  if (jsonConfig.options5.active == "True") {
    options5 = () => {
      return {
        host: jsonConfig.options5.host,
        port: jsonConfig.options5.port,
        database: jsonConfig.options5.database,
        user: jsonConfig.options5.user,
        password: jsonConfig.options5.password,
        role: jsonConfig.options5.role
      };
    };
    serverPort5 = () => {
      return jsonConfig.options5.srvPort;
    };
  }
  //Active Databases.
  activeDBs = () => {
    return {
      db1: jsonConfig.options1.active,
      db2: jsonConfig.options2.active,
      db3: jsonConfig.options3.active,
      db4: jsonConfig.options4.active,
      db5: jsonConfig.options5.active
    };
  };
}

module.exports = {
  options1: options1,
  serverPort1: serverPort1,
  options2: options2,
  serverPort2: serverPort2,
  options3: options3,
  serverPort3: serverPort3,
  options4: options4,
  serverPort4: serverPort4,
  options5: options5,
  serverPort5: serverPort5,
  flagMode: flagMode,
  activeDBs: activeDBs
};
